sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_read.h>
+
+#include <az_digit.h>
+
+#include <math.h>
+
+inline bool az_json_is_white_space(char const c) {
+  switch (c) {
+    case ' ': case '\t': case '\n': case '\r':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_is_esc(char const c) {
+  switch (c) {
+    case '\\':
+    case '"':
+    case '/':
+    case 'b':
+    case 'f':
+    case 'n':
+    case 'r':
+    case 't':
+      return true;
+  }
+  return false;
+}
+
+inline bool az_json_stack_is_empty(az_json_state const *const p) {
+  return p->stack == 1;
+}
+
+inline az_json_stack_item az_json_stack_last(az_json_state const *const p) {
+  return p->stack & 1;
+}
+
+az_json_state az_json_state_create(az_const_str const buffer) {
+  return (az_json_state){
+    .buffer = buffer,
+    .i = 0,
+    .stack = 1,
+  };
+}
+
+az_result az_json_get_char(az_const_str const buffer, size_t const i, char *const out_char) {
+  if (i == buffer.size) {
+    return AZ_JSON_ERROR_UNEXPECTED_END;
+  }
+  *out_char = az_const_str_item(buffer, i);
```

Comment:
AZ_ASSERT(out_char != NULL)

Reply:
That is a precondition, it should not be a random assert. I suggest we start using the PRECONDITION validation macros.

Reply:
Those should be put at the very top. Before checking for run-time errors. Reminder: precondition violation => a bug in a caller code.